active/0021-transfer-files-over-mqtt.md

```diff
@@ -0,0 +1,175 @@
+
+## AWS IoT MQTT-based file delivery (reference design)
+
+As an example of existing implementation we can look at AWS IoT Core [which provides functionality](https://docs.aws.amazon.com/iot/latest/developerguide/mqtt-based-file-delivery.html) to [deliver files to IoT devices](https://docs.aws.amazon.com/iot/latest/developerguide/mqtt-based-file-delivery-in-devices.html):
+
+![AWS IoT MQTT-based file delivery Device --> Cloud](0021-assets/aws-mqtt-file-delivery.png)
+
+## Design
+
+In the first phase we are going to focus on limited scope:
+
+* device-to-cloud file transfer only
+* reusing existing TCP connection
+* push only, i.e. broker cannot request file transfer, but device can start sending files any time
+* device may not know total file size when starting the transfer
+
+### Requirements
+
+* Based on the MQTT protocol. Use MQTT messages to encapsulate file bodies
+* Transport layer TCP, TCP/TLS and QUIC (TCP fallback) multiplexing connections with other service topics​
+* High concurrency (multiple uploads in progress)
+* MQTT Client single/bi-directional transfer (active file upload, passive file download)
+* Fault tolerance: disconnected retransmission (sender side)
+* Fragmentation: (multi parts, fragments)
+* Multi-channel: multiple TCP connections, QUIC multi-streams
+* Cancellation: Both parties can pause or cancel the transfer, and an event notification is required when the transfer is complete. (How to notify? What is the target of the notification?)
+* Minimal code changes on Client side, SDK support available. (low code/no code)
+* Minimize data retransmission, and reduce network pressure and bandwidth usage.
+* RestAPI interface Query transfer status
+* EMQX local incomplete file cache
+* Client API should be easy to use: easy to migrate from existing applications (ftp, nfs, s3)
+* File body Encryption and decryption (with external key management)?
+
+### Out of scope
+
+* Non-reliable transmission is not supported
+* No support for one-to-many transfers
+* No distributed cache storage
+* Broker does not aware of interrupted transfer, It will be client-side implementation
+
+### Limitations
+
+* File transfer is PUSH mechanism but not support POLL, the receiving side cannot request to initiate the transfer
+decisions of the size of a slice cannot be as natural as streaming.
+* Due to network bandwidth limitations, it may happen that multiple files are transferred in parallel but all fail to complete the transfer in the desired time.
+* Do not get too hung up on the length of the topic and do not take advantage of "topic alias" (MQTT 5.0)
+* Performance may be limited by disk IO (network in speed > disk out speed)
+* Overwrite existing file SEG
+* Linux filename length 255 bytes. path_max 4096
+* EMQX terminates file transfers, file message MQTT forwarding (as proxy) is not required.
+* QUIC in EMQX does not support HTTP/3 protocol, old client supports HTTP upload, how to smooth migration?
```

Comment:
We want to have a clear boundary/scope for this project.
Move the uncertain parts to "Discussion"
```suggestion
* When using MQTT over QUIC as transport, how to take advantage of QUIC protocol (or HTTP3) is not out of the scope of this EIP.
```